postgresql: always build with JIT enabled

wolfgangwalther opened the pull request:
Inspired by #385859.

This changes the build to always enable JIT - but to only enable it at run-time, when required. This keeps the runtime closure small without JIT, but allows enabling it without a rebuild. We can do this, because JIT is actually built as a shared module, which is loaded at run-time. We put it into a `-jit` output and only link it into the environment when requested.

Under the hood, this uses `withPackages` and adds the "JIT package" - thus, to be able to use `withPackages` on top of that, we also need to be able to apply `withPackages` repeatedly.

This cuts down the number of NixOS tests in half, because we don't need to run it for every version with and without JIT anymore. There really is no point in running everything with `llvmjit.so` in place, when the queries are not making use of it anyway.

Also, we only need to build each extension once and not twice, further reducing the number of rebuilds required for PRs touching `postgresql`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
postgresql: always build with JIT enabled

This changes the build to always enable JIT - but to only enable it at
run-time, when required. This keeps the runtime closure small without
JIT, but allows enabling it without a rebuild. We can do this, because
JIT is actually built as a shared module, which is loaded at run-time.
We put it into a -jit output and only link it into the environment when
requested.

Under the hood, this uses withPackages and adds the "JIT package" -
thus, to be able to use withPackages on top of that, we also need to be
able to apply withPackages repeatedly.

This cuts down the number of NixOS tests in half, because we don't need
to run it for every version with and without JIT anymore. There really
is no point in running everything with llvmjit.so in place, when the
queries are not making use of it anyway.

Also, we only need to build each extension once and not twice, further
reducing the number of rebuilds required for PRs touching postgresql.

wolfgangwalther committed:
postgresqlJitPackages.plv8: unbreak

This was originally marked as broken, because plv8 hardcoded gcc in its
Makefile. This has been resolved upstream in the meantime, so it builds
fine.

wolfgangwalther committed:
postgresqlPackages.plv8: mark as broken on darwin

Darwin seems to be principially supported according to upstream's issue
tracker[1], but it fails to build. Thus marking as broken instead of
unsupported.

[1]: https://github.com/plv8/plv8/issues/303#issuecomment-424883342

Ma27 commented:
I'd need a few days to take a look here.

My gut feeling was "meh" at first, but it's probably unjustified by now since it's probably based on the fact that I packaged JIT support a few years ago and made it its own package because of the obscene closure bloat that would've caused. Anyways, I'd like to think about this for a bit first before this lands.

wolfgangwalther committed:
kore: fix build

wolfgangwalther committed:
kore: use libpq 17

wolfgangwalther committed:
python3Packages.asyncpg: fix build

Ma27 commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -409,6 +409,9 @@ let
 
           # Stop lib depending on the -dev output of llvm
           remove-references-to -t ${llvmPackages.llvm.dev} "$out/lib/llvmjit${dlSuffix}"
```
This doesn't evaluate for at least static.
Maybe we should turn the entire jit stuff off in this case? Doesn't work anyways then.

Ma27 reviewed (commented):
Other than that, sgtm. Thanks!

wolfgangwalther committed:
pkgsStatic.{libpq,postgresql}: fix linking against libpq.a

This can be observed when building
haskellPackages.postgresql-libpq-pkgconfig. This is currently pinned to
postgresql_16 in pkgsStatic, but when unpinning it, it will fail with
the mentioned error on master. With this fix, it successfully builds
with postgresql_17 / libpq.

wolfgangwalther committed:
pkgsMusl.postgresql_jit: disable tests

Ma27 reviewed (approved):
Tests are fine and rest looks good to me.

Ma27 commented:
Ah! You merged already.
